Implement GraphicsLayerAsyncContentsDisplayDelegate for GraphicsLayerCARemote

mattwoodrow opened the pull request:
#### ac18fb1c144afb42706b9012347ce37fc47226da
<pre>
Implement GraphicsLayerAsyncContentsDisplayDelegate for GraphicsLayerCARemote
<a href="https://bugs.webkit.org/show_bug.cgi?id=249116">https://bugs.webkit.org/show_bug.cgi?id=249116</a>

Reviewed by Simon Fraser.

Relatively simple, it attempts to convert the ImageBuffer to a shareable bitmap handle, and sends
that directly to the UI process main thread if so.

* Source/WebCore/platform/graphics/ImageBuffer.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.messages.in:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::asyncSetLayerContents):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:
(WebKit::RemoteLayerTreeHost::asyncSetLayerContents):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/GraphicsLayerCARemote.cpp:
(WebKit::GraphicsLayerCARemoteAsyncContentsDisplayDelegate::GraphicsLayerCARemoteAsyncContentsDisplayDelegate):
(WebKit::GraphicsLayerCARemote::createAsyncContentsDisplayDelegate):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/GraphicsLayerCARemote.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeContext.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeContext.mm:
(WebKit::RemoteLayerTreeContext::drawingAreaIdentifier const):

Canonical link: <a href="https://commits.webkit.org/258289@main">https://commits.webkit.org/258289@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/05d9c4a694c92e85ff3bd06068070f632e5e6c29

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/101432 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/10590 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/34486 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/110702 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/170960 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/105412 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/11541 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/1442 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/93820 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/108519 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/107214 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/8781 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/92023 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/35306 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/90680 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/23431 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/78306 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/4192 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/24942 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/4254 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/1359 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/10341 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/44430 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5699 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/6017 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
```diff
@@ -554,39 +554,44 @@
     }
 }
 
+RetainPtr<id> RemoteLayerBackingStore::layerContentsBufferFromBackendHandle(ImageBufferBackendHandle& backendHandle, LayerContentsType contentsType)
```
`const ImageBufferBackendHandle& backendHandle`

smfr commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/GraphicsLayerCARemote.cpp:
```diff
@@ -90,4 +93,62 @@ void GraphicsLayerCARemote::moveToContext(RemoteLayerTreeContext& context)
+    }
+
+private:
+    Ref<IPC::Connection> m_connection;
```
Holding onto an `IPC::Connection` makes me a little nervous. Can this just call `WebProcess::singleton().parentProcessConnection()` each time?

mattwoodrow replied:
What makes you nervous about holding it?

My concern was that `WebProcess:terminate` could run concurrently with our worker thread trying to retrieve the connection, and we could end up with a dead object.

With the above code, the main-thread can terminate and disconnect the connection, and the worker will just safely drop the message.

That said, I'm fairly sure that the worker thread should have been shutdown by the time we're terminating the WebProcess, but it's hard to prove.

smfr commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/GraphicsLayerCARemote.cpp:
```diff
@@ -90,4 +93,62 @@ void GraphicsLayerCARemote::moveToContext(RemoteLayerTreeContext& context)
+        if (!backend)
+            return false;
+
+        buffer.flushDrawingContext();
```
Weird for `hasShareableBackend()` to have the side effect of flushing.

smfr commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/GraphicsLayerCARemote.cpp:
```diff
@@ -90,4 +93,62 @@ void GraphicsLayerCARemote::moveToContext(RemoteLayerTreeContext& context)
+        if (!backend)
+            return false;
+
+        clone->flushDrawingContext();
```
The ASSERT above had the side-effect of flushing, but you flush here again. Side effects are bad!

smfr commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/GraphicsLayerCARemote.cpp:
```diff
@@ -90,4 +93,62 @@ void GraphicsLayerCARemote::moveToContext(RemoteLayerTreeContext& context)
+        clone->flushDrawingContext();
+
+        auto* sharing = backend->toBackendSharing();
+        if (!is<ImageBufferBackendHandleSharing>(sharing))
```
This is what `hasShareableBackend()` did. Maybe just don't bother checking `hasShareableBackend()` at the top of the function, and just clone and check here.

mattwoodrow replied:
The idea was to try check if the original buffer could be shareable, before committing to the cost of cloning (and then verifying that the clone is also shareable).

smfr commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/GraphicsLayerCARemote.cpp:
```diff
@@ -90,4 +93,62 @@ void GraphicsLayerCARemote::moveToContext(RemoteLayerTreeContext& context)
+        if (!is<ImageBufferBackendHandleSharing>(sharing))
+            return false;
+
+        auto backendHandle = downcast<ImageBufferBackendHandleSharing>(*sharing).createBackendHandle(SharedMemory::Protection::ReadOnly);
```
Rather than the `is<>` above followed by the `downcast<>` here, just do it with one `dynamicDowncast<>`.

mattwoodrow committed:
Implement GraphicsLayerAsyncContentsDisplayDelegate for GraphicsLayerCARemote
https://bugs.webkit.org/show_bug.cgi?id=249116

Reviewed by Simon Fraser.

Relatively simple, it attempts to convert the ImageBuffer to a shareable bitmap handle, and sends
that directly to the UI process main thread if so.

* Source/WebCore/platform/graphics/ImageBuffer.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.messages.in:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::asyncSetLayerContents):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:
(WebKit::RemoteLayerTreeHost::asyncSetLayerContents):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/GraphicsLayerCARemote.cpp:
(WebKit::GraphicsLayerCARemoteAsyncContentsDisplayDelegate::GraphicsLayerCARemoteAsyncContentsDisplayDelegate):
(WebKit::GraphicsLayerCARemote::createAsyncContentsDisplayDelegate):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/GraphicsLayerCARemote.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeContext.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeContext.mm:
(WebKit::RemoteLayerTreeContext::drawingAreaIdentifier const):

Canonical link: https://commits.webkit.org/258289@main
